### PR TITLE
Toml/config cleanup

### PR DESCRIPTION
**Pull request by gustawlippa**
This PR addresses [MIM-1177](https://erlangsolutions.atlassian.net/browse/MIM-1177).

What has been done and why.
Maybe controversial:
- Removed description of the `mod_cowboy` listener. It was difficult to describe it clearly with the new TOML config and it's options are not parsed/validated in the new format. As the system metrics data shows, it was not used at all in the last year. From what I understand this functionality could also be configured by passing this module as a handler in `listen.http`?
- Removed `mod_revproxy` page from the modules list. The reasoning is similar as for the `mod_cowboy`. It seems to not be used, and with the current description of the configuration the module should not work (with TOML at least).
- Removed the `backend` option for `mod_bosh` as `"mnesia"` was the only valid option. Now it works like `mod_keystore` for example, where the option will not be parsed, validated, documented, but the module code may work with a new backend in the future easily as it doesn't change (only issue there is that the module relies on the `gen_mod` providing the correct default - `mnesia`. But such a dependency already existed in MIM in `mod_keystore`).
- Removed the `backend` option for `mod_inbox` for the same reasons.
- Replaced the `bounce` option with `bounce.enabled` for `mod_global_distrib`. It used to be a "boolean with only `false` being a valid option" which I found very confusing. This was because when someone wanted to have `bounce` enabled `bounce` became a TOML table as opposed to a key. Now there is a switch in the `bounce` section for this behaviour which keeps the behaviour of having bounce enabled by default.
- Replaced the `tls` option with `tls.enabled` for `mod_global_distrib` for the same reason. The only issue here is as `tls` is disabled by default (it needs some options in the `tls` section to be set), the "no section - disabled" approach seems more natural. Just for the consistency, it's changed to be similar to the `bounce` section in this regard.

Rather not controversial:
- Added `toml` to code blocks in with the configuration so that the syntax highlighting is enabled and consistent.
- Fixed an example configuration in `mod_keystore.md`
- Fixed small issues in the documentation like typos

### CODE REVIEW

**Comment by mongoose-im** [comment]
[8636.1](https://travis-ci.org/esl/MongooseIM/jobs/734991568) / Erlang 23.0.3 / small_tests / e7c4a690875f3f9af66654660f592d441638d2ca
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8636/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8636/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-12_12.27.16/lib.mongooseim.logs/run.2020-10-12_12.27.16/suite.log.html)

**Comment by mongoose-im** [comment]
[8637.1](https://travis-ci.org/esl/MongooseIM/jobs/735017484) / Erlang 23.0.3 / small_tests / cfb16e762a07f2bbf3b601cddd8399646e421839
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8637/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8637/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-12_14.06.13/lib.mongooseim.logs/run.2020-10-12_14.06.14/suite.log.html)

**Comment by mongoose-im** [comment]
[8638.1](https://travis-ci.org/esl/MongooseIM/jobs/735025206) / Erlang 23.0.3 / small_tests / 4404ebb961b6016f7d9809780abf9b073349c5fb
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8638/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8638/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-12_14.48.24/lib.mongooseim.logs/run.2020-10-12_14.48.25/suite.log.html)

**Comment by mongoose-im** [comment]
[8639.1](https://travis-ci.org/esl/MongooseIM/jobs/735350479) / Erlang 23.0.3 / small_tests / 70bbbf827fc5de8eb1ab35a8da8456c7c64e91fe
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8639/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8639/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-13_11.17.50/lib.mongooseim.logs/run.2020-10-13_11.17.51/suite.log.html)

**Comment by mongoose-im** [comment]
[8640.1](https://travis-ci.org/esl/MongooseIM/jobs/735362902) / Erlang 23.0.3 / small_tests / cd367129dbc0d1699eb9132e1465553e48919dd2
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8640/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8640/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-13_12.07.00/lib.mongooseim.logs/run.2020-10-13_12.07.00/suite.log.html)

File: doc/modules/mod_event_pusher.md [hunk 1]
```diff
@@ -45,4 +45,4 @@ Refer to their specific documentation to learn more about their functions and co
 [sns]: ./mod_event_pusher_sns.md
 [push]: ./mod_event_pusher_push.md
 [http_notification]: ./mod_event_pusher_http.md
-[rabbit]: ./mod_event_pusher_rabbit.md
\ No newline at end of file
+[rabbit]: ./mod_event_pusher_rabbit.md
```
Comment (chrzaszcz) [inline doc/modules/mod_event_pusher.md:48]: What is the difference here?

**Review by chrzaszcz — COMMENTED** [review]
Could you take care of the `mod_revproxy`section in `Modules.md` as well?

Maybe we could put a note there about `mod_revproxy` being deprecated and configurable only with the `cfg` format?

File: doc/modules/mod_event_pusher.md (hunk 1 above)
Comment (gustawlippa) [reply doc/modules/mod_event_pusher.md:48]: It's a new line added at the end of file, I don't know why GH fails to show it clearly.

**Comment by mongoose-im** [comment]
[8641.1](https://travis-ci.org/esl/MongooseIM/jobs/735388434) / Erlang 23.0.3 / small_tests / 582baa5b94886cabe7f5800f094ea961e6d3019f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8641/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8641/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-13_13.29.28/lib.mongooseim.logs/run.2020-10-13_13.29.28/suite.log.html)

**Comment by mongoose-im** [comment]
[8643.1](https://travis-ci.org/esl/MongooseIM/jobs/735703248) / Erlang 23.0.3 / small_tests / 2021b6433ff919860c5223ca1732df895b3a197a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8643/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8643/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-14_13.03.27/lib.mongooseim.logs/run.2020-10-14_13.03.27/suite.log.html)

**Comment by mongoose-im** [comment]
[8649.1](https://travis-ci.org/esl/MongooseIM/jobs/736323776) / Erlang 23.0.3 / small_tests / b77074f80e55ab5995dd7d016857bd9dedb3a959
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8649/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8649/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-16_10.42.05/lib.mongooseim.logs/run.2020-10-16_10.42.06/suite.log.html)

**Comment by mongoose-im** [comment]
[8654.1](https://travis-ci.org/esl/MongooseIM/jobs/737037292) / Erlang 23.0.3 / small_tests / 2128027240d0ffc13f5a3621d46c53df10251005
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8654/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8654/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-19_12.16.41/lib.mongooseim.logs/run.2020-10-19_12.16.41/suite.log.html)

**Comment by mongoose-im** [comment]
[8660.1](https://travis-ci.org/esl/MongooseIM/jobs/737396834) / Erlang 23.0.3 / small_tests / 92664e76bf99253ce7a29d4bf5ad529269f24a90
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8660/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8660/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-20_13.00.58/lib.mongooseim.logs/run.2020-10-20_13.00.58/suite.log.html)

**Comment by mongoose-im** [comment]
[8661.1](https://travis-ci.org/esl/MongooseIM/jobs/737397174) / Erlang 23.0.3 / small_tests / 585ab45b513a05077591436ac3704106a448bb83
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2909/8661/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2909/8661/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-10-20_13.48.29/lib.mongooseim.logs/run.2020-10-20_13.48.29/suite.log.html)

File: src/config/mongoose_config_parser_toml.erl [hunk 2]
```diff
@@ -1259,7 +1258,16 @@ mod_global_distrib_redis([<<"refresh_after">>|_], V) ->
...
-    [{max_retries, V}].
+    [{max_retries, V}];
+mod_global_distrib_bounce([<<"enabled">>|_], V) ->
+    [{enabled, V}].
+
+-spec format_global_distrib_bounce([option()]) -> option().
+format_global_distrib_bounce(Opts) ->
+    case proplists:lookup(enabled, Opts) of
+        {enabled, false} -> false;
+        _ -> Opts
```
Comment (chrzaszcz) [inline src/config/mongoose_config_parser_toml.erl:1269]: Let's remove `enabled` from the resulting opts: 
```suggestion
        _ -> maps:remove(enabled, Opts)
```

File: src/config/mongoose_config_parser_toml.erl [hunk 3]
```diff
@@ -1231,12 +1225,17 @@ mod_global_distrib_connections([<<"endpoint_refresh_interval_when_empty">>|_], V
...
 mod_global_distrib_connections([<<"tls">>|_] = Path, V) ->
-    TLSOpts = parse_section(Path, V),
+    TLSOpts = parse_section(Path, V, fun format_global_distrib_tls/1),
     [{tls_opts, TLSOpts}].
 
+-spec format_global_distrib_tls([option()]) -> option().
+format_global_distrib_tls(Opts) ->
+    case proplists:lookup(enabled, Opts) of
+        {enabled, false} -> false;
+        _ -> Opts
```
Comment (chrzaszcz) [inline src/config/mongoose_config_parser_toml.erl:1236]: IMO it should be `false` by default, so that setting the opts with no `enabled = true` would not enable them.
```suggestion
        {enabled, true} -> maps:remove(enabled, Opts);
        _ -> false
```

Also: removed `enabled` to clean up the resulting opts.

**Review by chrzaszcz — COMMENTED** [review]
Thanks for all the fixes.
I have some doubts about the implemented logic for `enabled`, I suggested corrections. Maybe it would be possible to add a test case clause for the corner cases as well?

File: src/config/mongoose_config_parser_toml.erl (hunk 3 above)
Comment (gustawlippa) [reply src/config/mongoose_config_parser_toml.erl:1236]: I figured somebody may omit `enabled` and provide other options, basically keeping the config as is right now. But it probably makes more sense as you suggest if we introduce the `enabled` option.

File: src/config/mongoose_config_parser_toml.erl [hunk 4]
```diff
@@ -1232,8 +1232,8 @@ mod_global_distrib_connections([<<"tls">>|_] = Path, V) ->
 -spec format_global_distrib_tls([option()]) -> option().
 format_global_distrib_tls(Opts) ->
     case proplists:lookup(enabled, Opts) of
-        {enabled, false} -> false;
-        _ -> Opts
+        {enabled, true} -> maps:remove(enabled, Opts);
```
Comment (chrzaszcz) [inline src/config/mongoose_config_parser_toml.erl:1235]: OMG it's a proplist, not a map!
Needs a correction with a force push :-D

Comment (gustawlippa) [reply src/config/mongoose_config_parser_toml.erl:1235]: Yeah I just noticed it failed, I'll push it with some changes to the `.md` file as well because the example would be wrong right now.

**Review by chrzaszcz — APPROVED** [review]
